### PR TITLE
Fixed Material Builder To Avoid Reading Material Type Source File

### PR DESCRIPTION
**Pull request by santorac**
These changes ensure that Material Builder does not read the .materialtype source file when processing .material files, which fixes https://github.com/o3de/o3de/issues/6594. This issue occured because the the builder was reading source .materialtype files but did not declare a source dependency, which is incorrect. 

# Details

Before, when the .material files were processed the JSON serializer would load and use the .materialtype source file to figure out the expected data type for each property. The JSON system is very flexible and would convert bools, floats, ints, etc. automatically to the appropriate type indicated in the property layout. So if you reprocess a .material when a property data type changes, then the builder would convert the old data types for you and the final MaterialAsset would work fine. But if you update the .materialtype _without_ reprocessing the .material file, then the data types wouldn't match at runtime and the material fails to load.

So we have two choices: either (1) declare the source dependency on the the .materialtype source file (so any change to the .materialtype will force the .material to reprocess and see the new type), or (2) don't read the .materialtype file at all. We shouldn't add the source dependency because it will introduce time-consuming rebuilds of every .material file when the associated .materialtype changes (that was the whole point of https://github.com/o3de/o3de/pull/6982). So instead we change the material JSON serializers to avoid opening the .materialtype file. In that case, there is no way to know what data type the property layout is expecting, so all we can do is load the data as it appears in the .material file. Then the MaterialAsset::Finalize function does any necessary data type conversion.

This has the added benefit of simplifying the code by removing the need for a custom JSON serializer for the MaterialSourceData class.

Also updated all .material files to have materialTypeVersion instead of propertyLayoutVersion. This was renamed in code at some point but we forgot to rename in the files. Before this was silently ignored but since I removed MaterialSourceDataSerializer, this started being reported as a warning.

# Testing

RPI unit tests pass.
Opened a bunch of materials in the material editor and made sure they loaded correctly. (This wasn't working for inherited materials in an earlier iteration of my changes, but it's fine now).
ASV dx12 passes (other than one known issue, no change from before).
Confirmed https://github.com/o3de/o3de/issues/6594 is fixed
Opened a couple AutomatedTesting levels in the Editor.
Deleted and rebuild the asset cache, and re-ran all the above tests.

### CODE REVIEW

File: Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp [hunk 1]
```diff
@@ -109,6 +109,75 @@ namespace AZ
             return m_wasPreFinalized;
         }
 
+        template<typename T>
```
Comment (jeremyong-az) [inline Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp:112]: My eyes might be tired, but why does this need to be a template function if the returned type is also a variant? Internally it looks like we're chaining the `T` cast right after the rtti-based cast but does the second cast do anything? (Probably should just be referring to `MaterialPropertyValue` to understand this better, but just noting that this is an open question as the reader for the time being)

**Review by jeremyong-az — APPROVED** [review]

File: Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp (hunk 1 above)
Comment (santorac) [reply Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp:112]: MaterialPropertyValue is a AZStd::variant, which does not support any kind of casting. If it's an int, the only valid call to variant::GetValue is GetValue<int>(). This function tries to convert the given MaterialPropertyValue to a different type of MaterialPropertyValue, according to the template parameter T.

File: Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp [hunk 2]
```diff
@@ -109,6 +109,81 @@ namespace AZ
...
+            {
+                return value;
+            }
+        }
+        
+        //! Attempts to convert an AZ::Vector[2-4] MaterialPropertyValue to another AZ::Vector[2-4] type @T.
+        //! Any extra elements will be dropped or set to 0.0 as needed.
+        //! If the original MaterialPropertyValue is not a Vector type, the original value is returned.
+        template<typename VectorT>
+        MaterialPropertyValue CastVectorMaterialPropertyValue(const MaterialPropertyValue& value)
```
Comment (gadams3) [inline Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp:146]: Should this support converting a single floating point value to a vector or vice versa?

**Review by gadams3 — APPROVED** [review]
This is great. Removing any custom serializers we can will make things so much simpler.

File: Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp (hunk 2 above)
Comment (santorac) [reply Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp:146]: My goal was just to replace whatever automatic conversions work in the JSON serialization system, and that doesn't do conversion between float and Vector types.

**Pull request merged**